Sources/NIO/Codec.swift

```diff
@@ -101,115 +129,265 @@ private extension ChannelHandlerContext {
+extension B2MDBuffer {
+    /// Start processing some bytes if possible, if we receive a returned buffer (through `.available(ByteBuffer)`)
+    /// we _must_ indicate the processing has finished by calling `finishProcessing`.
+    mutating func startProcessing() -> BufferAvailability {
+        switch self.state {
+        case .processingInProgress:
+            return .bufferAlreadyBeingProcessed
+        case .ready where self.buffers.count > 0:
+            var buffer = self.buffers.removeFirst()
+            buffer.writeBuffers(self.buffers)
+            self.buffers.removeAll(keepingCapacity: true)
+            if buffer.readableBytes > 0 {
+                self.state = .processingInProgress
+                return .available(buffer)
+            } else {
+                return .nothingAvailable
             }
+        case .ready:
+            assert(self.buffers.count == 0)
+            return .nothingAvailable
         }
     }
 
-    /// Calls `decode` until there is nothing left to decode.
-    public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
-        // Either merge the received data into the existing cumulationBuffer or use it as the cumulationBuffer if none exists yet.
-        if self.cumulationBuffer != nil {
-            var buffer = self.unwrapInboundIn(data)
-            self.cumulationBuffer!.write(buffer: &buffer)
+
+    mutating func finishProcessing(remainder buffer: ByteBuffer) -> Void {
+        assert(self.state == .processingInProgress)
+        self.state = .ready
+        if buffer.readableBytes > 0 {
+            self.buffers.prepend(buffer)
         } else {
-            self.cumulationBuffer = self.unwrapInboundIn(data)
+            var buffer = buffer
+            buffer.clear()
+            buffer.writeBuffers(self.buffers)
+            self.buffers.removeAll(keepingCapacity: true)
+            self.buffers.append(buffer)
         }
+    }
 
-        ctx.withThrowingToFireErrorAndClose {
-            try self.decodeLoop(ctx: ctx, decodeFunc: self.decode)
+    mutating func append(buffer: ByteBuffer) {
+        if buffer.readableBytes > 0 {
+            self.buffers.append(buffer)
         }
+    }
+}
 
-        // Discard the cumulationBuffer or discard read bytes if needed.
-        guard let buffer = self.cumulationBuffer, buffer.readableBytes > 0 else {
-            self.cumulationBuffer = nil
+// MARK: B2MDBuffer Helpers
+private extension ByteBuffer {
+    mutating func writeBuffers(_ buffers: CircularBuffer<ByteBuffer>) {
+        guard buffers.count > 0 else {
             return
         }
-
-        // Check if we should reclaim some bytes and if so do it.
-        if self.shouldReclaimBytes(buffer: self.cumulationBuffer!) {
-            self.cumulationBuffer!.discardReadBytes()
+        var allReadableBytes: Int = 0
+        for buffer in buffers {
+            allReadableBytes += buffer.readableBytes
+        }
+        self.reserveCapacity(self.writerIndex + allReadableBytes)
```

Comment:
nit: you could use `var allReadableBytes = self.writerIndex` before to eliminate on `+` operation ;)